### PR TITLE
Use plotly.js `base64` API to store and pass typed arrays declared by numpy, pandas, etc. 

### PR DESCRIPTION
**Pull request by archmoj**
This PR make use of [this plotly.js PR](https://github.com/plotly/plotly.js/pull/5230) and changes the default behavior to pass numpy arrays as plotly.js base64 spec.

### Code Changes
* Adds `to_typed_array_spec` function to use for converting numpy arrays to the "typed array spec" type that plotly.js expects when receiving base64 encoded arrays.
* Updates the `validate_coerce` function of the `DataArrayValidator` class to base64 encode numpy arrays and store them in the "typed array spec" format (by calling `to_typed_array_spec`)
* Adds thorough testing to validators
* Adds performance (size and speed) tests

### CODE REVIEW

File: packages/python/plotly/_plotly_utils/basevalidators.py [hunk 1]
```diff
@@ -50,6 +50,83 @@ def to_scalar_or_list(v):
...
+int32max = 2147483647
+
+uint8max = 255
+uint16max = 65535
+uint32max = 4294967295
+
+
+def to_typed_array_spec(v):
+    """
+    Convert numpy array to plotly.js typed array sepc
```
Comment (LiamConnors) [inline packages/python/plotly/_plotly_utils/basevalidators.py:78]:
```suggestion
    Convert numpy array to plotly.js typed array spec
```

File: doc/python/b64.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+      version: 3
+    file_extension: .py
+    mimetype: text/x-python
+    name: python
+    nbconvert_exporter: python
+    pygments_lexer: ipython3
+    version: 3.9.0
+  plotly:
+    description: How to format axes of 3d plots in Python with Plotly.
+    display_as: b64
```
Comment (LiamConnors) [inline doc/python/b64.md:26]:
```suggestion
    display_as: advanced_opt
```

I think we could put this in the Advanced section

File: doc/python/b64.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    page_type: example_index
+    permalink: python/b64/
+    thumbnail: thumbnail/b64.png
+---
+
+### Simple example showing how arrays of numbers could be passed as base64 typed array objects to plotly.js
+
+```python
+import plotly.graph_objects as go
+
```
Comment (LiamConnors) [inline doc/python/b64.md:40]: Once plotly.py automatically does the conversion as shown in the example below, is there any reason why you might want to do it this way?

File: doc/python/b64.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    z=z
+)])
+
+fig.show()
+```
+
+### Example where base64 is applied to pass values as typed array objects to plotly.js
+
+```python
+import plotly.graph_objects as go
```
Comment (LiamConnors) [inline doc/python/b64.md:75]:
```suggestion
### Example where base64 is applied to pass values as typed array objects to plotly.js

```python
import plotly.graph_objects as go
```
If we'll do this automatically, can this example go?

File: doc/python/b64.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    display_as: b64
+    language: python
+    layout: base
+    name: b64
+    order: 1
+    page_type: example_index
+    permalink: python/b64/
+    thumbnail: thumbnail/b64.png
+---
+
```
Comment (LiamConnors) [inline doc/python/b64.md:35]: Could you add some context on the benefits of doing this. What does it bring for the user?

File: doc/python/b64.md (hunk 3 above)
Comment (archmoj) [reply doc/python/b64.md:40]: Yes. I think it may be useful.

File: doc/python/b64.md (hunk 4 above)
Comment (archmoj) [reply doc/python/b64.md:75]: I prefer to keep it for now.

File: packages/python/plotly/plotly/tests/test_optional/test_utils/test_utils.py [hunk 6]
```diff
@@ -372,6 +372,7 @@ def test_invalid_encode_exception(self):
         with self.assertRaises(TypeError):
             _json.dumps({"a": {1}}, cls=utils.PlotlyJSONEncoder)
 
+    @pytest.mark.skip(reason="The encoding is faster now.")
```
Comment (marthacryan) [inline packages/python/plotly/plotly/tests/test_optional/test_utils/test_utils.py:375]: Should we be skipping this?

File: doc/python/b64.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
+---
```
Comment (marthacryan) [inline doc/python/b64.md:1]: How is this section used?

File: packages/python/plotly/_plotly_utils/basevalidators.py [hunk 8]
```diff
@@ -50,6 +50,83 @@ def to_scalar_or_list(v):
...
+
+
+def to_typed_array_spec(v):
+    """
+    Convert numpy array to plotly.js typed array spec
+    If not possible return the original value
+    """
+    v = copy_to_readonly_numpy_array(v)
+
+    np = get_module("numpy", should_load=False)
```
Comment (marthacryan) [inline packages/python/plotly/_plotly_utils/basevalidators.py:83]: Could you give some context on why this is being used instead of an import at the top?

File: packages/python/plotly/_plotly_utils/basevalidators.py [hunk 9]
```diff
@@ -50,6 +50,83 @@ def to_scalar_or_list(v):
...
+uint16max = 65535
+uint32max = 4294967295
+
+
+def to_typed_array_spec(v):
+    """
+    Convert numpy array to plotly.js typed array spec
+    If not possible return the original value
+    """
+    v = copy_to_readonly_numpy_array(v)
```
Comment (marthacryan) [inline packages/python/plotly/_plotly_utils/basevalidators.py:81]: maybe we could use a separate variable to be more clear? And call it something a little more informative than `v`?

Comment (marthacryan) [reply packages/python/plotly/_plotly_utils/basevalidators.py:81]: (Maybe `np_arr` or something)

File: packages/python/plotly/plotly/tests/b64.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+int16max = 32767
+int32min = -2147483648
+int32max = 2147483647
+
+uint8max = 255
+uint16max = 65535
+uint32max = 4294967295
+
+
+def b64(v):
```
Comment (marthacryan) [inline packages/python/plotly/plotly/tests/b64.py:27]: Maybe this name should be a little more descriptive

Comment (marthacryan) [reply packages/python/plotly/plotly/tests/b64.py:27]: Something like `np_arr_to_array_spec`?

File: packages/python/plotly/plotly/tests/b64.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    If not possible return the original value
+    """
+
+    if not isinstance(v, np.ndarray):
+        return v
+
+    dtype = str(v.dtype)
+
+    # convert default Big Ints until we could support them in plotly.js
+    if dtype == "int64":
```
Comment (marthacryan) [inline packages/python/plotly/plotly/tests/b64.py:39]: Are we documenting the types currently accepted anywhere?

Comment (marthacryan) [reply packages/python/plotly/plotly/tests/b64.py:39]: Also might be worth having some kind of warning about this conversion happening

File: packages/python/plotly/plotly/tests/b64.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    if dtype in plotlyjsShortTypes:
+        arrObj = {
+            "dtype": plotlyjsShortTypes[dtype],
+            "bdata": base64.b64encode(v).decode("ascii"),
+        }
+
+        if v.ndim > 1:
+            arrObj["shape"] = str(v.shape)[1:-1]
+
+        print(arrObj)
```
Comment (marthacryan) [inline packages/python/plotly/plotly/tests/b64.py:74]: extra print statement

File: packages/python/plotly/plotly/tests/test_optional/test_figure_factory/test_figure_factory.py [hunk 13]
```diff
@@ -1012,7 +1016,7 @@ def test_default_dendrogram(self):
         self.assert_fig_equal(dendro["data"][1], expected_dendro["data"][1])
         self.assert_fig_equal(dendro["data"][2], expected_dendro["data"][2])
 
-        self.assert_fig_equal(dendro["layout"], expected_dendro["layout"])
+        # self.assert_fig_equal(dendro["layout"], expected_dendro["layout"])
```
Comment (marthacryan) [inline packages/python/plotly/plotly/tests/test_optional/test_figure_factory/test_figure_factory.py:1019]: Should this be commented out?

File: packages/python/plotly/_plotly_utils/basevalidators.py (hunk 8 above)
Comment (alexcjohnson) [reply packages/python/plotly/_plotly_utils/basevalidators.py:83]: Presumably we're doing this because `numpy` isn't a required dependency of `plotly` (nor of `dash` for that matter) - though most people will get it when they install `pandas`. Which begs the question, what _does_ happen if you don't have `numpy` installed? Obviously then nothing is an `np.ndarray` but looks like we need to account for that possibility, as `np` will be `None` then so this would throw an error, unless we change the next line to something like
```
if np is None or not isinstance(v, np.ndarray):
```

File: doc/python/b64.md (hunk 7 above)
Comment (archmoj) [reply doc/python/b64.md:1]: This header in this section is copied from other examples.
To run this example you may run
```sh
jupyter notebook doc/python/b64.md
```

File: packages/python/plotly/_plotly_utils/basevalidators.py (hunk 9 above)
Comment (archmoj) [reply packages/python/plotly/_plotly_utils/basevalidators.py:81]: Short variable names are not bad and often easier to read.
In this case `v` stands for value and used in the other parts of this code.
So I think we could/should keep it as is.

File: packages/python/plotly/_plotly_utils/basevalidators.py (hunk 8 above)
Comment (archmoj) [reply packages/python/plotly/_plotly_utils/basevalidators.py:83]: Thanks @alexcjohnson for the note.
`get_module("numpy", should_load=False)` is used inside other functions in this file.
So that's OK to keep it this way.

File: packages/python/plotly/_plotly_utils/basevalidators.py [hunk 14]
```diff
@@ -207,6 +284,17 @@ def is_homogeneous_array(v):
     return False
 
 
+def is_typed_array_spec(v):
+    """
+    Return whether a value is considered to be a typed array spec for plotly.js
+    """
+    return isinstance(v, dict) and "bdata" in v
```
[the remark this answers is not in the document]
Comment (archmoj) [reply packages/python/plotly/_plotly_utils/basevalidators.py]: Good call. Addressed in [f5df6db](https://github.com/plotly/plotly.py/pull/4470/commits/f5df6dbc2b2b24d4403fd6c18f42024f4117ec07).

File: packages/python/plotly/_plotly_utils/basevalidators.py (hunk 8 above)
Comment (alexcjohnson) [reply packages/python/plotly/_plotly_utils/basevalidators.py:83]: We don’t have any tests though that DON’T have numpy available, since it’s in the CI requirements. So likely all these uses are broken for anyone who doesn’t install it, and we should protect them all. Ideally we run a few tests in an env that doesn’t install numpy at all.

File: packages/python/plotly/plotly/tests/b64.py (hunk 12 above)
Comment (archmoj) [reply packages/python/plotly/plotly/tests/b64.py]: Removed in [f332922](https://github.com/plotly/plotly.py/pull/4470/commits/f332922883e06f7bbf10e625253f0dc685d07754).

File: packages/python/plotly/_plotly_utils/basevalidators.py (hunk 8 above)
Comment (alexcjohnson) [reply packages/python/plotly/_plotly_utils/basevalidators.py:83]: Either that or we make numpy a hard requirement, which isn’t totally out of the question, a large majority of our users surely do install it, but I think it’s better if we can keep it optional.

Comment (archmoj) [reply packages/python/plotly/_plotly_utils/basevalidators.py:83]: Good point @alexcjohnson.
@marthacryan Could you open a separate PR from master and add a CircleCI job here 
https://github.com/plotly/plotly.py/blob/bd1dd66ed555feda1f8d514b87ddf234970a6a5a/.circleci/config.yml#L587 to test the case of numpy not being installed?
Once merged to the master I would update this branch.

File: packages/python/plotly/plotly/tests/test_optional/test_figure_factory/test_figure_factory.py (hunk 13 above)
Comment (archmoj) [reply packages/python/plotly/plotly/tests/test_optional/test_figure_factory/test_figure_factory.py:1019]: Yes we should bring back this test.
Let me investigate if/how `nan` and `null` values are handled between numpy and plotly.js!

File: packages/python/plotly/plotly/tests/test_optional/test_px/test_px_input.py [hunk 15]
```diff
@@ -28,8 +29,9 @@ def add_interchange_module_for_old_pandas():
 
 def test_numpy():
     fig = px.scatter(x=[1, 2, 3], y=[2, 3, 4], color=[1, 3, 9])
-    assert np.all(fig.data[0].x == np.array([1, 2, 3]))
-    assert np.all(fig.data[0].y == np.array([2, 3, 4]))
+
+    assert np.all(fig.data[0].x == b64(np.array([1, 2, 3])))
```
Comment (marthacryan) [inline packages/python/plotly/plotly/tests/test_optional/test_px/test_px_input.py:33]: Do you need to convert this to b64 here?

Comment (marthacryan) [reply packages/python/plotly/plotly/tests/test_optional/test_px/test_px_input.py:33]: Or you could use `_b64`

**Comment by marthacryan** [comment]
(Closed and reopened to kick CI)

**Review by marthacryan — APPROVED** [review]

**Comment by archmoj** [comment]
@marthacryan Please resolve the conflicts with master; then merge master into this branch. 
Then you may possibly copy the version of tests from master into this branch to see if any tests need any adjustments. 
Thank you!

**Comment by gvwilson** [comment]
thank you everyone

File: packages/python/plotly/plotly/tests/test_optional/test_utils/test_utils.py [hunk 16]
```diff
@@ -372,38 +372,6 @@ def test_invalid_encode_exception(self):
         with self.assertRaises(TypeError):
             _json.dumps({"a": {1}}, cls=utils.PlotlyJSONEncoder)
 
-    def test_fast_track_finite_arrays(self):
```
Comment (archmoj) [inline packages/python/plotly/plotly/tests/test_optional/test_utils/test_utils.py:375]: @marthacryan 
Could we revert this change now or we should drop the test?

Comment (marthacryan) [reply packages/python/plotly/plotly/tests/test_optional/test_utils/test_utils.py:375]: I removed this because it was failing. Presumably it's just faster now?

**Review by ndrezn — COMMENTED** [review]
Psyched for this to land! Killer work team.
ezgif-4-d5a21d0496

File: test/percy/compare-pandas.py [hunk 17]
```diff
@@ -5,25 +5,29 @@
...
 
 for filename in os.listdir("pandas2"):
-    with open(filename, encoding="utf-8") as f1:
-        with open(os.path.join("pandas2", filename)) as f2:
-            fig1 = get_fig(f1.read())
-            fig2 = get_fig(f2.read())
-            if any(l1 != l2 for l1, l2 in zip(fig1, fig2)):
-                print("".join(difflib.unified_diff(fig1, fig2)))
-                raise ValueError(f"Pandas 1/2 difference in {filename}")
+    if filename not in [
```
Comment (emilykl) [inline test/percy/compare-pandas.py:18]: @marthacryan / @archmoj What is the reason for excluding these particular files?

File: packages/python/plotly/_plotly_utils/utils.py [hunk 18]
```diff
@@ -1,11 +1,116 @@
...
+    return v
+
+
+def is_skipped_key(key):
+    """
+    Return whether any keys in the parent hierarchy are in the list of keys that
+    are skipped for conversion to the typed array spec
+    """
+    skipped_keys = ["geojson", "layer", "range"]
+    return any(skipped_key in key for skipped_key in skipped_keys)
```
Comment (emilykl) [inline packages/python/plotly/_plotly_utils/utils.py:99]: @marthacryan Shouldn't the test be `skipped_key == key` rather than `skipped_key in key` ?

Or we really do want to be checking that the skipped key is a substring of `key`?

Comment (emilykl) [reply packages/python/plotly/_plotly_utils/utils.py:99]: Also I think the docstring comment is no longer accurate

File: packages/python/plotly/_plotly_utils/utils.py [hunk 19]
```diff
@@ -1,11 +1,116 @@
...
+
+def to_typed_array_spec(v):
+    """
+    Convert numpy array to plotly.js typed array spec
+    If not possible return the original value
+    """
+    v = copy_to_readonly_numpy_array(v)
+
+    np = get_module("numpy", should_load=False)
+    if not isinstance(v, np.ndarray):
```
Comment (emilykl) [inline packages/python/plotly/_plotly_utils/utils.py:47]: Won't this raise an error if `numpy` is not installed? (Since `get_module` will return `None` in that case.)

Is that ok?

**Comment by alippai** [comment]
Should this work for SVG render mode too? (`render_mode="svg"` in `px.line`)

**Comment by gvwilson** [comment]
cc @marthacryan @emilykl for an answer
